#!/usr/bin/env Rscript

# The Germline Genomics of Cancer (G2C)
# Copyright (c) 2025-Present, Ryan L. Collins and the Dana-Farber Cancer Institute
# Contact: Ryan Collins <Ryan_Collins@dfci.harvard.edu>
# Distributed under the terms of the GNU GPL v2.0

# Simulate null LD values for a set of variants based on their AFs


#########
# Setup #
#########
# Load necessary libraries and constants
options(scipen=1000, stringsAsFactors=F)
require(argparse, quietly=TRUE)
require(RLCtools, quietly=TRUE)


#############
# Functions #
#############
# LD simulation procedure for a single variant
sim.ld <- function(af, n, pos, max.r2=1, max.attempts=1000, digits=6){
  # If AF << 1 / 2*N, return perfect LD even despite max.r2
  if(af < 1 / (200 * n) | (1 - af) < 1 / (200 * n)){return(1)}
  sim.r2 <- 1
  attempts <- 0
  while(sim.r2 > max.r2){
    v1 <- simulate.gts(af, n, seed=pos+attempts, return.dosage=T)
    v2 <- simulate.gts(af, n, seed=pos+1+attempts, return.dosage=T)
    new.r2 <- cor(v1, v2)^2
    if(!is.na(new.r2)){sim.r2 <- round(new.r2, digits)}
    attempts <- attempts + 1
    if(attempts >= max.attempts){
      return(sim.r2)
    }
  }
  return(sim.r2)
}


###########
# RScript #
###########
# Parse command line arguments and opions
parser <- ArgumentParser(description="Simulate null LD values for a set of variants")
parser$add_argument("--bed", metavar=".bed", type="character", required=TRUE,
                    help="Input .bed of site metrics generated by CollectSiteMetrics")
parser$add_argument("-N", "--sample-size", metavar="integer", type="numeric",
                    help=paste("Number of sample genotypes to simulate for each",
                               "variant. If not specified, will infer approximate",
                               "sample size from AF and AC in --bed"))
parser$add_argument("--max-r2", metavar="float", type="numeric", default=0.05,
                    help=paste("Simulated R2 values equal to or above this",
                               "threshold will be rerun [default: 0.05]"))
parser$add_argument("--out-tsv", metavar="path", type="character",
                    help="Path to output .tsv", default="./vcf_qc")
args <- parser$parse_args()

# # DEV:
# args <- list("bed" = "~/scratch/site_bench_inputs_v2_may14/dfci-g2c.v1.initial_qc.all_svs.bed.gz",
#              "sample_size" = 2000,
#              "max_r2" = 0.05,
#              "out_tsv" = "~/scratch/ld_sim_test.tsv")

# Read input BED file with variants to simulate
x <- read.table(args$bed, header=T, check.names=F, sep="\t", quote="", comment.char="")

# Approximate number of samples with non-null genotypes for each variant
if(!is.null(args$sample_size)){
  x$N <- args$sample_size
}else{
  x$N <- round(x$ac / (2 * x$af))
}

# Simulate LD R2 for all variants
x$r2 <- sapply(1:nrow(x), function(i){
  sim.ld(af=x$af[i], n=x$N[i], pos=x$start[i], max.r2=args$max_r2)
})

# Append simulated LD R2 pairs to file
write.table(x[, c("vid", "r2")], args$out_tsv, append=T, col.names=F,
            row.names=F, sep="\t", quote=F)
